src/promptflow-azure/promptflow/azure/_utils/_tracing.py

```diff
@@ -2,33 +2,68 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # ---------------------------------------------------------
 
+import time
 import typing
 
 from azure.ai.ml import MLClient
 from azure.core.exceptions import ResourceNotFoundError
 from azure.identity import AzureCliCredential, DefaultAzureCredential
 
 from promptflow._constants import CosmosDBContainerName
-from promptflow._sdk._tracing import get_ws_tracing_base_url
 from promptflow._sdk._utils import extract_workspace_triad_from_trace_provider
+from promptflow._utils.logger_utils import get_cli_sdk_logger
 from promptflow.azure import PFClient
 from promptflow.azure._restclient.flow_service_caller import FlowRequestException
 from promptflow.exceptions import ErrorTarget, UserErrorException
 
+_logger = get_cli_sdk_logger()
+
+COSMOS_INIT_POLL_TIMEOUT_SECOND = 600  # 10 minutes
+COSMOS_INIT_POLL_INTERVAL_SECOND = 10  # 10 seconds
+
 
 def _get_credential() -> typing.Union[AzureCliCredential, DefaultAzureCredential]:
     try:
         credential = AzureCliCredential()
         credential.get_token("https://management.azure.com/.default")
         return credential
-    except Exception:
+    except Exception:  # pylint: disable=broad-except
         return DefaultAzureCredential()
 
 
 def _create_trace_provider_value_user_error(message: str) -> UserErrorException:
     return UserErrorException(message=message, target=ErrorTarget.CONTROL_PLANE_SDK)
 
 
+def _init_workspace_cosmos_db(pf_client: PFClient) -> None:
+    # SDK will call PFS async API to execute workspace Cosmos DB initialization
+    # and poll the status until it's done, the signal is the response is not None
+    start_time = time.time()
+    while True:
+        try:
+            cosmos_res = pf_client._traces._init_cosmos_db()
+            if cosmos_res is not None:
+                return
+        except FlowRequestException:
+            # ignore request error and continue to poll in next iteration
+            pass
+        # set a timeout here to prevent the potential infinite loop
+        if int(time.time() - start_time) > COSMOS_INIT_POLL_TIMEOUT_SECOND:
+            break
+        prompt_msg = "The workspace Cosmos DB initialization is still in progress..."
+        _logger.info(prompt_msg)
+        time.sleep(COSMOS_INIT_POLL_INTERVAL_SECOND)
+    # initialization does not finish in time, we need to ensure the Cosmos resource is ready
+    # so print error log and raise error here
+    error_msg = (
+        "The workspace Cosmos DB initialization is still in progress "
+        f"after {COSMOS_INIT_POLL_TIMEOUT_SECOND} seconds, "
+        "please wait for a while and retry the command."
```

Comment:
remove " the command" as it might be a sdk experience

Reply:
removed